feat: Adds `defn-` and `def-` macros

TimDeve opened the pull request:
Adding these macros as a shortand for declaring a def or defn and making them `hidden` and `private`, useful to keep things internal to a module. Similar to Clojure's [`defn-`](https://clojuredocs.org/clojure.core/defn-).

TimDeve commented:
Thinking of adding some example of using `hidden` & `private` in the module section of the [Language Guide](https://github.com/carp-lang/Carp/blob/master/docs/LanguageGuide.md) and introduce these two macros at the same time. Does that sound approriate?

scolsen commented:
> Thinking of adding some example of using `hidden` & `private` in the module section of the [Language Guide](https://github.com/carp-lang/Carp/blob/master/docs/LanguageGuide.md) and introduce these two macros at the same time. Does that sound approriate?

Sounds great! We should add some tests for these too.

TimDeve commented:
Yeah I was thinking about the tests yesterday. Would you expect some "error message" tests?

hellerve commented:
> Yeah I was thinking about the tests yesterday. Would you expect some "error message" tests?

I’d say so! We can use the `test-for-errors` setup.

hellerve commented on core/Macros.carp:
```diff
@@ -133,12 +133,19 @@
 
-(defmodule Dynamic
-
-  (defndynamic quoted [x]
```
Where did this go?

TimDeve commented:
> I’d say so! We can use the test-for-errors setup.

Added the tests a few minutes before your message. Doing docs next.

TimDeve commented:
Took me weeks to find the time to make this simple PR but I think it's finally ready for review 😄

hellerve commented on docs/LanguageGuide.md:
```diff
@@ -227,6 +156,88 @@ These can only be used at the REPL and during macro evaluation. Here's a subset
 
+
+### Structs
+Any structure type defined in Carp has an init method that can be used to create a new instance. It must be called with all the arguments in the order they are defined.
```
Could we wrap `init` in backticks here?

Also, maybe "in the order in which they are defined" or "in definition order" might be better?

hellerve commented on docs/LanguageGuide.md:
````diff
@@ -227,6 +156,88 @@ These can only be used at the REPL and during macro evaluation. Here's a subset
+  (Right [b]))
+```
+
+A Variant can be created with the same syntax as call expression:
````
`A variant can be created through a function call`?

hellerve commented on docs/LanguageGuide.md:
````diff
@@ -227,6 +156,88 @@ These can only be used at the REPL and during macro evaluation. Here's a subset
+    _ (logic-other)))
+```
+
+Note that match works with *values* (not references) takes ownership over the value being matched on. If you instead want to match on a reference, you can use `match-ref`:
````
I think backticks around `match` would be good here, and that there is an `and` missing after the parentheses.

hellerve reviewed (commented):
I realized about halfway through that the docs I was annotating weren’t new but rearranged. My comments still hold, I think 😆

hellerve commented:
I think it would be good to use these forms in stdlib in a follow-up PR, since this could remove a lot of code and test the forms at the same time.

TimDeve commented:
The follow up PR sounds like a good idea. About the comments you've added I was thinking of making a refractor PR for the language guide, the style is not very consistent and some of it seems straight up wrong with the changes in the language since it was written, so I could put your suggested changes it that.

hellerve commented:
> The follow up PR sounds like a good idea. About the comments you've added I was thinking of making a refractor PR for the language guide, the style is not very consistent and some of it seems straight up wrong with the changes in the language since it was written, so I could put your suggested changes it that.

Sounds like a good plan! I’ll be happy to help and/or review :)

TimDeve commented:
Once/if this is merged I'll open a draft PR with your suggested changes and we can comment on it on what we add to that.

eriksvedang commented:
Looks good! The changes to the language guide is a bit hard to follow though, has anything been deleted or have things just moved around?

TimDeve commented:
Sorry the merged diff looks really messy, it's easier to follow if you go commit by commit. I moved the module and interface section after the struct section first, and then I added a section about `private` and `hidden` in the module section.

eriksvedang commented:
@TimDeve Alright, thanks I'll do that!

eriksvedang reviewed (approved):
@hellerve had some comments that you can choose to address, apart from that I think we're good to go!

eriksvedang commented:
@TimDeve Maybe you already discussed that and this should just go in as-is now?

TimDeve commented:
Yeah the comment where on existing docs so we're going to address them in a different PR.

eriksvedang commented:
Great work!
